### ISSUE TITLE
"#if\377" makes glslangValidator allocate resources in an infinite loop.

### ISSUE THREAD

**Issue by xorgy** [KhronosGroup/glslang#25]
`"#if\377"` triggers an issue in glslangValidator causing an infinite hot loop which also allocates memory until it crashes.

**Comment by xorgy** [comment]
@johnkslang Have fun. :-)

**Comment by xorgy** [comment]
This works for a variety of preprocessor directives, and in some states it loops infinitely without the resource leak(which could be worse, since the process may never be terminated).

**Comment by xorgy** [comment]
```
(gdb) step
scanToken (ppToken=<optimized out>, this=<optimized out>) at /home/aaron/src/glslang/glslang/MachineIndependent/preprocessor/PpContext.h:221
221             while (! inputStack.empty()) {
(gdb) step
glslang::TPpContext::extraTokenCheck (this=this@entry=0x7fffffffca30, atom=atom@entry=288, ppToken=ppToken@entry=0x7fffffff47b0, token=<optimized out>)
    at /home/aaron/src/glslang/glslang/MachineIndependent/preprocessor/Pp.cpp:340
340                 token = scanToken(ppToken);
(gdb) step
scanToken (ppToken=<optimized out>, this=<optimized out>) at /home/aaron/src/glslang/glslang/MachineIndependent/preprocessor/PpContext.h:221
221             while (! inputStack.empty()) {
(gdb) 
```
Seems to be bouncing between `extraTokenCheck` and `scanToken`

**Comment by johnkslang** [comment]
I'm not seeing the kind of issue you describe.  The #if gets a '\' token back and says that's a bad expression and everything is okay.  Was there more to the shader?  It might be connected to something else going on.

I did enhance #error, indirectly, with aead93aafd1b115ae7c67f5c1df8a90536c4421e by allowing '\' to be a real token, but that doesn't change that it has no real purpose and should generally cause a bad token to be noticed.

Note that GLSL/ESSL do not have escape sequences.  The string "#if\376" simply means the following token stream: hash, "if", backslash, 376, which is in fact what I see.  Since backslash is not a valid expression token, that error message is given.

**Comment by xorgy** [comment]
@johnkslang I mean the character that the octal number "\377" represents. I just can't put the literal character here because it isn't reproduced.

**Comment by xorgy** [comment]
@johnkslang I mean the character that the octal number "\377" represents. I just can't put the literal character here because it isn't reproduced.

**Comment by johnkslang** [comment]
Thanks, I reproduced the issues.  They are fixed in c777fc2c4c14991af0d98c094f532405364701f2, which generally made this space safer and more rational.

**Comment by xorgy** [comment]
Glad to hear it, thanks.
